### ISSUE TITLE
Is it possible to continue learning using traditional PPO after having learned through imitation?

### ISSUE THREAD

**Issue by jlanis** [Unity-Technologies/ml-agents#1032]
Is it possible to train an agent initially using imitation learning, and then have them continue to learn by having multiple agents play against one another?

**Comment by traveltrousers** [comment]
No reason why not... just resume the training referencing the trained brain.

**Comment by icaro56** [comment]
@traveltrousers , Are you sure about this?

Because the graphs generated by Tensorflow of PPO and BC are different.

**Comment by vincentpierre** [comment]
In the current version of ml-agents, there is no way to switch between imitation learning and ppo. That us because as @icaro56  said, the graphs are very different. Bootstrapping RL with IL is hard because the agent tends to forget very quickly his imitation training when switching to RL. There are other algorithms that can be used (inverse RL for instance) but we do not have a version of this in the current version.

**Comment by icaro56** [comment]
I'm trying to implement this. I was able to load the model trained by BC in the PPO, but it seems that happens what @vincentpierre said.

**_Bootstrapping RL with IL is hard because the agent tends to forget very quickly his imitation training when switching to RL_**

I will test with a simpler version of my problem to make sure the learning is being forgotten.

**Comment by mbaske** [comment]
@jlanis An alternative approach could be training an agent to mimick another agent's behavior using PPO. I had some success doing that for bootstrapping walking motions (continuous action space). Create a duplicate of your agent and hardcode its behavior, e.g. as a heuristic brain. The original agent receives its observations from the hardcoded one. Set a negative reward proportional to the summed difference between both agents' vector action values. Eventually, the trained actions should converge towards the hardcoded ones. Stop when they're close enough, then continue training your agent using its own observations and rewards.

**Comment by icaro56** [comment]
@mbaske , what observations do you get from hardcoded agent? Can you give us a example?

**Comment by icaro56** [comment]
Thank you @mbaske .

**Comment by icaro56** [comment]
@mbaske , 

My project is similar to Bomberman. In my case, if the student choose the wrong action then he will be in other position different from specialist. In this case, do you do what?

1) Set a negative reward and restart the episode to the start position in both agents (?)

2) Set a negative reward and to the next iteration set the student position to same specialist position (?)

**Comment by mbaske** [comment]
Hi @icaro56 - That sounds like you're dealing with a discrete action space. I'm not sure if the method I described works for discrete actions, haven't tried that yet.

Two more general points though: The student doesn't actually have to do anything while it is learning the hardcoded actions. As long as it is receiving its observations from another agent, there's no need for it to be active in the environment.
Also, there should be some variation in the hardcoded agent's observations. If the student's behavior is too tightly fitted to a specific sequence of observations, it will likely have a harder time when it continues training by itself later on.

**Comment by icaro56** [comment]
thank you again.

**Comment by icaro56** [comment]
After updating the ML-Agents plugin to version 0.5, I tried loading the trained model with BC to continue training with PPO (via --load flag). Now it looks like it worked, but the PPO agent does not try to improve what was learned, it just keeps repeating the movement learned with the CB.

I tried change the hiperparameters but it does not works.

**Comment by awjuliani** [comment]
Hi @icaro56 I am not sure how your method for using the BC model under PPO works. What were you changing before to allow the training to take place? Unfortunately it currently isn't supported to simply load a model trained with BC and use it with PPO. The neural network graphs created by the two algorithms are not compatible. I am currently working on revamping Imitation Learning though, and we plan to support doing this officially in the future.

**Comment by MarcoMeter** [comment]
What about if you let a BC trained agent gather experiences, which can be utilized by the PPO trainer to train a new model?

**Comment by icaro56** [comment]
@awjuliani ,

**Hi @icaro56 I am not sure how your method for using the BC model under PPO works. What were you changing before to allow the training to take place?** 

First, I change the --load code in the trainer_controller.py. I insert 2 booleans to control the customLoadMode and the pre trained model path location. The newLoadMode and continueTrainWithPPOAfterBC respectively control these.

```

            if self.load_model:
                self.logger.info('Loading Model...')
                ckpt = None
                if self.continueTrainWithPPOAfterBC == True:
                    ckpt = tf.train.get_checkpoint_state(self.model_path_imitation)
                    if ckpt is None:
                        self.logger.info('The model {0} could not be found. Make sure you specified the right '
                                         '--run-id'.format(self.model_path_imitation))
                else:
                    ckpt = tf.train.get_checkpoint_state(self.model_path)
                    if ckpt is None:
                        self.logger.info('The model {0} could not be found. Make sure you specified the right '
                                     '--run-id'.format(self.model_path))
                if (self.newLoadMode == True):
                    self.logger.info(load new version')
                    var_list = self.optimistic_restore_vars(ckpt.model_checkpoint_path) if ckpt != None else None

                    sess.run(init)

                    # saver used only for restore.
                    saverForRestore = tf.train.Saver(max_to_keep=self.keep_checkpoints, var_list=var_list)
                    saverForRestore.restore(sess, ckpt.model_checkpoint_path)
                else:
                    self.logger.info('load old version')
                    saver.restore(sess, ckpt.model_checkpoint_path)
            else:
                sess.run(init)


```

Almost any tensor graph that is used by BC is contained in the PPO model. Thinking from the point of view of the theory of mathematical sets, the PPO set contains the BC set. But, in Tensorflow, if we try load different models it will throw errors. For to solve this problem, I am using a function that restore only the tensors that exist in both models. In the code above, see optimistic_restore_vars function.

The optimistic_restore_vars function is used to restore only the existing tensors in both BC and PPO. In Tensorflow an error occurs at the time of loading if the current model has more tensors than the model to be loaded. So we started the current model by forcing the sess.run (init) command before loading so that the tensors that exist only in the PPO (current model) are initialized and with the command saverForRestore.restore (...) we restore the tensors which already existed in the previous model.
```

    def optimistic_restore_vars(self, model_checkpoint_path):
        reader = tf.train.NewCheckpointReader(model_checkpoint_path)
        saved_shapes = reader.get_variable_to_shape_map()
        var_names = sorted([(var.name, var.name.split(':')[0]) for var in tf.global_variables()
                            if var.name.split(':')[0] in saved_shapes])
        restore_vars = []
        name2var = dict(zip(map(lambda x: x.name.split(':')[0], tf.global_variables()), tf.global_variables()))
        with tf.variable_scope('', reuse=True):
            for var_name, saved_var_name in var_names:
                curr_var = name2var[saved_var_name]
                var_shape = curr_var.get_shape().as_list()
                if var_shape == saved_shapes[saved_var_name]:
                    restore_vars.append(curr_var)
        return restore_vars


```

I calculate the value_estimate and the entropy in my BC custom model too and add in the inference_dict_list. But calculating or not value_estimate and entropy, the behavior is being the same.

In the bc/Model.py discrete action space code, after self.actions_probs:
```

                value = tf.layers.dense(hidden, 1, activation=None)
                self.value = tf.identity(value, name="value_estimate")
                # self.entropy = -tf.reduce_sum(self.action_probs * tf.log(self.action_probs + 1e-10), axis=1)
                action_idx = [0] + list(np.cumsum(self.act_size))
                self.entropy = tf.reduce_sum((tf.stack([
                    tf.nn.softmax_cross_entropy_with_logits_v2(
                        labels=tf.nn.softmax(self.action_probs[:, action_idx[i]:action_idx[i + 1]]),
                        logits=self.action_probs[:, action_idx[i]:action_idx[i + 1]])
                    for i in range(len(self.act_size))], axis=1)), axis=1)


```

**Comment by icaro56** [comment]
@awjuliani , Did you try using my code above?

**Comment by ishaybe** [comment]
hi,
@awjuliani when do think you will add it to official version?

**Comment by awjuliani** [comment]
Hi @ishaybe 

We will be adding official support for this in the beginning of 2019. We will be combining the model code for both kinds of trainers, so that the two will be interchangeable.

**Comment by havenbrown1108** [comment]
Is there an update on this functionality?
